statsd/statsd.go

```diff
@@ -120,36 +141,103 @@ func (c *Client) watch() {
 }
 
 func (c *Client) append(cmd string) error {
-	c.Lock()
 	c.commands = append(c.commands, cmd)
 	// if we should flush, lets do it
 	if len(c.commands) == c.bufferLength {
 		if err := c.flush(); err != nil {
-			c.Unlock()
 			return err
 		}
 	}
-	c.Unlock()
 	return nil
 }
 
+func (c *Client) joinMaxSize(cmds []string, sep string, maxSize int) ([][]byte, []int) {
+	c.buffer.Reset() //clear buffer
+
+	var frames [][]byte
+	var ncmds []int
+	sepBytes := []byte(sep)
+
+	elem := 0
+	for _, cmd := range cmds {
+		needed := len(cmd)
+
+		//if a single command is too big, flush it and hope the UDP fragmented packet
+		//doesn't get dropped.
+		if needed > maxSize {
+			frames = append(frames, []byte(cmd))
+			ncmds = append(ncmds, 1)
+			continue
+		}
+
+		if elem != 0 {
+			needed = needed + len(sep)
+		}
+
+		if c.buffer.Len()+needed <= maxSize {
+			if elem != 0 {
+				c.buffer.Write(sepBytes)
+			}
+			c.buffer.WriteString(cmd)
+			elem++
+		} else {
+			tmpBuf := make([]byte, c.buffer.Len())
+			copy(tmpBuf, c.buffer.Bytes())
```

Comment:
Any reason why you don't want to use the String() method for the buffer? I guess efficiency, but we're ultimately doing something similar here with an allocation and a copy.